logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java

```diff
@@ -0,0 +1,468 @@
+        private static final Collection<JrubyEventExtLibrary.RubyEvent> EMPTY_RETURN =
+            Collections.emptyList();
+
+        /**
+         * Trivial {@link Dataset} that simply returns an empty collection of elements.
+         */
+        private static final Dataset EMPTY_DATASET = new Dataset() {
+
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown) {
+                return EMPTY_RETURN;
+            }
+
+            @Override
+            public void clear() {
+            }
+        };
+
+        private final Collection<Dataset> parents;
+
+        /**
+         * <p>Builds a terminal {@link Dataset} from the given parent {@link Dataset}s.</p>
+         * <p>If the given set of parent {@link Dataset} is empty the sum is defined as the
+         * trivial dataset that does not invoke any computation whatsoever.</p>
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean)} is always
+         * {@link Collections#emptyList()}.
+         * @param parents Parent {@link Dataset} to sum and terminate
+         * @return Dataset representing the sum of given parent {@link Dataset}
+         */
+        public static Dataset from(final Collection<Dataset> parents) {
+            final int count = parents.size();
+            final Dataset result;
+            if (count > 0) {
+                result = new Dataset.TerminalDataset(parents);
+            } else {
+                result = EMPTY_DATASET;
+            }
+            return result;
+        }
+
+        private TerminalDataset(final Collection<Dataset> parents) {
+            this.parents = parents;
+        }
+
+        @Override
+        public Collection<JrubyEventExtLibrary.RubyEvent> compute(final RubyIntegration.Batch batch,
+            final boolean flush, final boolean shutdown) {
+            parents.forEach(dataset -> dataset.compute(batch, flush, shutdown));
+            this.clear();
+            return EMPTY_RETURN;
+        }
+
+        @Override
+        public void clear() {
+            for (final Dataset parent : parents) {
+                parent.clear();
+            }
+        }
+
+    }
+
+    /**
+     * {@link Dataset} that results from the {@code if} branch of its backing
+     * {@link EventCondition} being applied to all of its dependencies.
+     */
+    final class SplitDataset implements Dataset {
+
+        private final Collection<Dataset> parents;
+
+        private final EventCondition func;
+
+        private final Collection<JrubyEventExtLibrary.RubyEvent> data;
+
+        private final Collection<JrubyEventExtLibrary.RubyEvent> complement;
+
+        private final Dataset opposite;
+
+        private boolean done;
+
+        public SplitDataset(final Collection<Dataset> parents, final EventCondition func) {
+            this.parents = parents;
+            this.func = func;
```

Comment:
can we rename this to `eventCondition`? I think that's more clear. This isn't a generic function.